### DISCUSSION TITLE
start qemu-system-arm in Eclipse, always time out

### DISCUSSION THREAD

**Discussion by spark952** [xpack-dev-tools/qemu-arm-xpack#21] [General]
Debug in Eclipse, I want to use qemu-systme-arm, instead of qemu-system-gnuarmeclipse.exe, to support more machines.
In Eclipse debug configureation page, debugger tab, the Actual Excutable is:
C:\Program Files\GNU ARM Eclipse\QEMU\xpack-qemu-arm-7.0.0-1\bin\qemu-system-arm.exe

But when click debug button, the Eclipse will always report the error as below:
An internal error occurred during: "Starting QEMU".
RequestMonitor: Sequence "", result for executing step #1 = Status ERROR: org.eclipse.cdt.dsf code=10005 Request for monitor: 'RequestMonitor (org.eclipse.embedcdt.debug.gdbjtag.core.dsf.GnuMcuGdbServerBackend$GdbServerStep$1@3b8ea542): Status OK: unknown code=0 OK' resulted in a rejected execution exception., done() method called more than once

In the same env,  qemu-system-gnuarmeclipse.exe worked well.
Is anything else should I do to use qemu-system-arm.exe here?
Thanks.

**Comment by ilg-ul** [comment]
Unfortunately the existing Eclipse QEMU plug-in is specific to `qemu-system-gnuarmeclipse` (which was customised for Eclipse), and cannot be used with `qemu-system-arm`.

I'm facing the same issues, and I plan to add a new plug-in to Eclipse, but it'll take some time.

The workaround is to start `qemu-systme-arm` separately, and disable **Start QEMU locally** in the plug-in.

For example, for a Cortex-M3 test I'm debugging right now, I use:

```
$ /Users/ilg/Library/xPacks/@xpack-dev-tools/qemu-arm/7.0.0-1.1/.content/bin/qemu-system-arm "--machine" "mps2-an385" "--nographic" "-d" "unimp,guest_errors" "--semihosting-config" "enable=on,target=native,arg=test,arg=one" -s -S
```

**Comment by spark952** [comment]
Thank you for your reply, @ilg-ul  :)
Would you please tell me where could I get some info about the changes between qemu-system-arm  and qemu-system-gnuarmeclipse version?  In other words, I`d like to know what did we do to customize qemu for Eclipse.
Thanks.

**Comment by ilg-ul** [comment]
There are many changes, hard to summarise.

`qemu-system-gnuarmeclipse` is built from:

- <https://github.com/xpack-dev-tools/qemu/tree/xpack-legacy>

while `qemu-system-arm` is built from:

- <https://github.com/xpack-dev-tools/qemu/tree/xpack>

which follows the upstream repo.

**Comment by spark952** [comment]
OK，understood，thank you :)
Another question, what`s the main difference between the queue-system-arm in xPack release and Official release？I saw the machine list and cpu list are both the same with each other.

**Reply by ilg-ul** [reply]
Currently there are no functional differences. If you do a git compare you should see only the greeting message updated to include _xPack_, since they refused my patch to configure it at build time.

**Reply by spark952** [reply]
OK，seems the qemu-system-arm in the xPack release almost the same with that in the Official release, am i right?
I tried the workarount you supplied, and it works and could help to keep my work going on ,thank you very much :)
And hope to see your new plugin to use qemu-system-arm in Eclipse.

**Reply by ilg-ul** [reply]
> the same with that in the Official release, am i right?

Yes, it is the same.

**Comment by ilg-ul** [comment]
FYI, I migrated my unit-tests to the new qemu-system-arm v7.x. I'm running the tests on Cortex-M0, Cortex-M7F, Cortex-A15 and Cortex-A72 (aarch64), using semihosting. 

The tricky parts were:

- the memory map (including the mechanism to initialise the heap and the stack in a separate region)
- the initialisations required in the reset handler

So far the experience seems to be positive; I'll try to find time for the new Eclipse plug-in.

**Reply by spark952** [reply]
Glad to know it :)
And I now can launch qemu-system-arm from Eclipse by ignoreing to check the stdout of gdb server, seems could be used to debug.
